### DISCUSSION TITLE
Would like help understanding Terragrunt behavior

### DISCUSSION THREAD

**Discussion by nicholasarvelo** [gruntwork-io/knowledge-base#702] [Help]
The Terragrunt code I am sharing does what I want it to do. However, I am not 100% certain I understand why it is working. It doesn't align with what I understand from reading Gruntwork.io's documentation. I was hoping someone could help me better understand.

This is my directory structure:
```
terragrunt/
├── development/
│   ├── ecs/
│   │   ├── ecs_service_name/
│   │   │   ├── data.tf
│   │   │   ├── main.tf
│   │   │   ├── task_definition.tf
│   │   │   ├── terragrunt.hcl
│   │   │   └── variables.tf
│   └── environment_variables.hcl
└── backend_configuration.hcl
```

The contents of `terragrunt.hcl`:
```
include "backend_configuration" {
  path = find_in_parent_folders("backend_configuration.hcl")
}

include "environment_variables" {
  path = find_in_parent_folders("environment_variables.hcl")
}
```

The contents of `environment_variables.hcl`:
```
locals {
  environment = basename(dirname(find_in_parent_folders("environment_variables.hcl")))
}

inputs = {
  environment = local.environment
}
```

The contents of `backend_configuration.hcl`:
```
remote_state {
  backend = "s3"

  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }

  config = {
    bucket         = "terraform-bucket-name"
    key            = "${path_relative_to_include()}/terraform.tfstate"
    dynamodb_table = "dynamodb-table-name"
    region         = "us-east-1"
  }
}
```

As I already mentioned above, everything is working. Meaning:

- `terragrunt.hcl` adopts the configuration details from the two parent Terragrunt configuration files.
- `backend.tf` is dynamically created with the expected key value.
- `environment` evaluates to `development` and is exported for Terraform interpolation.

So, this is my question:

Per Gruntwork's documentation:

> find_in_parent_folders() searches up the directory tree from the current terragrunt.hcl file and returns the absolute path to the first terragrunt.hcl in a parent folder

> The function takes an optional name parameter that allows you to specify a different filename to search for

This would lead me to believe that in order for `environment` to evaluate to `development` , I should be setting
```
environment = basename(dirname(find_in_parent_folders("environment_variables.hcl")))
```
to
```
environment = basename(dirname(find_in_parent_folders("backend_configuration.hcl")))
```
However this is not the resulting behavior. Setting it this way evaluates to `terragrunt`.

It only generates the desired result if set to:
```
environment = basename(dirname(find_in_parent_folders("environment_variables.hcl")))
```
or, interestingly, if set to:
```
environment = basename(dirname(find_in_parent_folders()))
```

I would expect to fail with the latter example since there is no `terragrunt.hcl`. 

Can someone help me understand why it is behaving as such?

---

<ins datetime="2023-04-27T15:10:42Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110128">Tracked in ticket #110128</a></p>
</ins>

**Comment by denis256** [comment]
Hi,
since `backend_configuration.hcl` is located in `terragrunt`, `find_in_parent_folders()` will return path to `terragrunt` directory, not to `development`

In my tests:
```
$ tree 
discussion-702/
├── backend_configuration.hcl
└── development
    ├── ecs
    │   └── ecs_service_name
    │       ├── file.txt
    │       ├── main.tf
    │       ├── terraform.tfstate
    │       ├── terraform.tfstate.backup
    │       └── terragrunt.hcl
    └── environment_variables.hcl

# backend_configuration.hcl
find_in_parent_folders("backend_configuration.hcl") = /projects/gruntwork/terragrunt-tests/discussion-702/backend_configuration.hcl
dirname(find_in_parent_folders("backend_configuration.hcl")) = /projects/gruntwork/terragrunt-tests/discussion-702
basename(dirname(find_in_parent_folders("backend_configuration.hcl"))) = discussion-702

# environment_variables.hcl
find_in_parent_folders("environment_variables.hcl") = /projects/gruntwork/terragrunt-tests/discussion-702/development/environment_variables.hcl
dirname(find_in_parent_folders("environment_variables.hcl")) = /projects/gruntwork/terragrunt-tests/discussion-702/development
basename(dirname(find_in_parent_folders("environment_variables.hcl"))) = development
```
References:
https://developer.hashicorp.com/terraform/language/functions/dirname
https://developer.hashicorp.com/terraform/language/functions/basename
https://github.com/denis256/terragrunt-tests/tree/master/discussion-702

**Reply by nicholasarvelo** [reply]
Thanks @denis256,

But I don't have `environment_variables.hcl` looking for `backend_configuration.hcl`. In my `environment_variables.hcl` I have `find_in_parent_folders("environment_variables.hcl")` set, which somehow works. Isn't that command as written essentially telling `environment_variables.hcl` to look for itself? It doesn't make sense to me.

**Reply by denis256** [reply]
It depends on where `find_in_parent_folders()` is executed - in the current case, after includes resolving, `find_in_parent_folders("environment_variables.hcl")` is executed in `development/ecs/ecs_service_name` and it is finding `environment_variables.hcl`
